python/lsst/obs/base/defineVisits.py

```diff
@@ -655,12 +742,121 @@ def group(self, exposures: List[DimensionRecord]) -> Iterable[VisitDefinitionDat
             yield VisitDefinitionData(
-                instrument=instrument, id=visitId, name=visitName, exposures=exposuresInGroup
+                instrument=instrument,
+                id=visitId,
+                name=visitName,
+                exposures=exposuresInGroup,
+                visit_system_mask=2**self.config.visitSystemId,
             )
 
-    def getVisitSystem(self) -> Tuple[int, str]:
+    def getVisitSystems(self) -> List[Tuple[int, str]]:
+        # Docstring inherited from GroupExposuresTask.
+        return [(self.config.visitSystemId, self.config.visitSystemName)]
+
+
+class _GroupExposuresByCounterAndExposuresConfig(GroupExposuresConfig):
+    visitSystemId = Field(
+        doc="Integer ID of the visit_system implemented by this grouping algorithm.",
+        dtype=int,
+        default=5,  # Not appliccable. One-to-one ID OR by-group-counter ID.
+    )
+    visitSystemName = Field(
+        doc="String name of the visit_system implemented by this grouping algorithm.",
+        dtype=str,
+        default="by-counter-and-exposures",  # Not real name.
+    )
+
+
+@registerConfigurable("one-to-one-and-by-counter", GroupExposuresTask.registry)
+class _GroupExposuresByCounterAndExposuresTask(GroupExposuresTask, metaclass=ABCMeta):
+    """An exposure grouping algorithm that uses the sequence start and
+    sequence end metadata but also creates one-to-one
+
+    This algorithm uses the exposure.seq_start and
+    exposure.seq_end fields to collect related snaps.
+    It also groups single exposures.
+    """
+
+    ConfigClass = _GroupExposuresByCounterAndExposuresConfig
+    _visit_systems = {"one-to-one": 0, "by-group-counter": 2}
+
+    def group(self, exposures: List[DimensionRecord]) -> Iterable[VisitDefinitionData]:
+        # Docstring inherited from GroupExposuresTask.
+        groups = defaultdict(list)
+        for exposure in exposures:
+            groups[exposure.day_obs, exposure.seq_start, exposure.seq_end].append(exposure)
+        for visit_key, exposures_in_group in groups.items():
+            instrument = exposures_in_group[0].instrument
+
+            # It is possible that the first exposure in a visit has not
+            # been ingested. This can be determined and if that is the case
+            # we can not reliably define the multi-exposure visit.
+            skip_multi = False
+            sorted_exposures = sorted(exposures_in_group, key=lambda e: e.seq_num)
+            first = sorted_exposures.pop(0)
+            if first.seq_num != first.seq_start:
+                # Special case seq_num == 0 since that implies that the
+                # instrument has no counters and therefore no multi-exposure
+                # visits.
+                if first.seq_num == 0:
+                    self.log.warning(
+                        "First exposure for visit %s is not present. Skipping the multi-snap definition.",
+                        visit_key,
+                    )
```

Comment:
Seems like it'd be safer and simpler to raise, unless you have a shouldn't-be-fatal way this could happen in mind.

Reply:
We can't raise. You can encounter this problem if you have had a transfer problem such that the second exposure has been ingested but the first has not. Issuing a warning but defining the one-to-one visit seems better than stopping everything.

Reply:
Of course.  No idea how I missed that the first time around.